Porting FxCop Rules to Roslyn.md

```diff
@@ -0,0 +1,66 @@
+
+In the remainder of this document, we explain the principles we used to decide how to factor the new Roslyn-based analyzers, enumerate the specific NuGet packages into which the analyzers will be factored, and describe in a little more detail how we decided which FxCop rules to port.
+
+## Factoring principles
+
+* In the spirit of [Code-Aware libraries](https://channel9.msdn.com/Events/Build/2015/3-725), if a rule is about the usage of a specific API, and the rule doesn't make sense if that API is not referenced, then that rule should ship with that API. For example, rules about `ImmutableArray` (which resides in System.Collections.Immutable.dll) should reside in an analyzer assembly System.Collections.Immutable.Analyzers.dll, which would be included in the System.Collections.Immutable NuGet package.
+
+* Some types reside in different .NET assemblies, depending on which flavor of .NET you use. For example, in the .NET Framework, `IDisposable` resides in mscorlib.dll, whereas in [.NET Core](http://blogs.msdn.com/b/dotnet/archive/2014/11/12/net-core-is-open-source.aspx),    it resides in System.Runtime.dll. Where should we place analyzers that examine uses of `IDisposable`: in mscorlib.Analyzers.dll or in System.Runtime.Analyzers.dll? We should choose the .NET Core version of the types; that is, we should place the `IDisposable` analyzers in System.Runtime.Analyzers.dll.
+
+    The rationale for this choice is that developers using .NET Core, which is delivered as a set of NuGet packages, will automatically get exactly the API-specific analyzers they need. Developers using .NET Framework will still need to manually download the API-specific analyzers. For those developers, we might consider creating a consolidated NuGet package containing the analyzers for all types in the .NET framework. By doing these two things, we minimize the number of times developers have to search for and download API-specific analyzer packages. 
+
+* Rules that do not relate to the usage of specific APIs, but relate instead to more general coding guidelines, should be organized according to the intended purpose of those guidelines. For example, some rules might help API authors produce consistent public APIs, but those rules might not make sense for test assemblies. (We will package those analyzers in Microsoft.ApiDesignGuidelines.Analyzers.dll, which by the way is essentially the original purpose of FxCop.) As another example, there might be some rules that restrict the expressiveness of the language (by discouraging the use of certain language features) in order to gain a performance advantage. Such rules would only apply in a specific context where that tradeoff is acceptable, and hence it would be useful to place them in a separate NuGet package.
```

Comment:
Not strictly so. More arguably uninteresting history: the .NET design guideline enforcement was mostly embodied by a tool named 'UrtCop' (for universal runtime cop). 'FxCop' was rebranded when it was officially picked up by an internal development team, the Fx (for 'framework's) was intended to point to its expanded purposes to flag incorrect/problematic usage patterns across the whole of the .NET framework (including System.Windows.Forms, System.Data, System.Xml, etc). So, FxCop was always intended for this dual purpose (API surface area scan + Fx usage). In practice, it turned out that bundling the design guidelines with the correctness became more problematic, as analysis generated API design results for code that didn't strictly require fixes. This wasn't obvious in the early days of FxCop, however, because everyone was new to the platform and had a vested interest in understanding the design guidelines. There was significant willingness/interest, therefore, in getting design guideline feedback and fixing related issues, for all code. In fact, we rec'd requests to enable the design guidelines for all private types/methods as well from some eager users (this facility still exists today). Hard to imagine these requests occurring today. :)